README
NAME
    Devel::REPL - a modern perl interactive shell

SYNOPSIS
      my $repl = Devel::REPL->new;
      $repl->load_plugin($_) for qw(History LexEnv);
      $repl->run

    Alternatively, use the 're.pl' script installed with the distribution

      system$ re.pl

DESCRIPTION
    This is an interactive shell for Perl, commonly known as a REPL - Read,
    Evaluate, Print, Loop. The shell provides for rapid development or
    testing of code without the need to create a temporary source code file.

    Through a plugin system, many features are available on demand. You can
    also tailor the environment through the use of profiles and run control
    files, for example to pre-load certain Perl modules when working on a
    particular project.

USAGE
    To start a shell, follow one of the examples in the "SYNOPSIS" above.

    Once running, the shell accepts and will attempt to execute any code
    given. If the code executes successfully you'll be shown the result,
    otherwise an error message will be returned. Here are a few examples:

     $_ print "Hello, world!\n"
     Hello, world!
     1
     $_ nosuchfunction
     Compile error: Bareword "nosuchfunction" not allowed while "strict subs" in use at (eval 130) line 5.

     $_

    In the first example above you see the output of the command ("Hello,
    world!"), if any, and then the return value of the statement (1).
    Following that example, an error is returned when the execution of some
    code fails.

    Note that the lack of semicolon on the end is not a mistake - the code
    is run inside a Block structure (to protect the REPL in case the code
    blows up), which means a single statement doesn't require the semicolon.
    You can add one if you like, though.

    If you followed the first example in the "SYNOPSIS" above, you'll have
    the History and LexEnv plugins loaded (and there are many more
    available). Although the shell might support "up-arrow" history, the
    History plugin adds "bang" history to that so you can re-execute chosen
    commands (with e.g. "!53"). The LexEnv plugin ensures that lexical
    variables declared with the "my" keyword will automatically persist
    between statements executed in the REPL shell.

    When you "use" any Perl module, the "import()" will work as expected -
    the exported functions from that module are available for immediate use:

     $_ carp "I'm dieeeing!\n"
     String found where operator expected at (eval 129) line 5, near "carp "I'm dieeeing!\n""
             (Do you need to predeclare carp?)
     Compile error: syntax error at (eval 129) line 5, near "carp "I'm dieeeing!\n""
     BEGIN not safe after errors--compilation aborted at (eval 129) line 5.

     $_ use Carp

     $_ carp "I'm dieeeing!\n"
     I'm dieeeing!
      at /usr/share/perl5/Lexical/Persistence.pm line 327
     1
     $_

    To quit from the shell, hit "Ctrl+D" or "Ctrl+C".

      MSWin32 NOTE: control keys won't work if TERM=dumb
      because readline functionality will be disabled.

  Run Control Files
    For particular projects you might well end up running the same commands
    each time the REPL shell starts up - loading Perl modules, setting
    configuration, and so on. A run control file lets you have this done
    automatically, and you can have multiple files for different projects.

    By default the "re.pl" program looks for "$HOME/.re.pl/repl.rc", and
    runs whatever code is in there as if you had entered it at the REPL
    shell yourself.

    To set a new run control file that's also in that directory, pass it as
    a filename like so:

     system$ re.pl --rcfile myproject.pc

    If the filename happens to contain a forwardslash, then it's used
    absolutely, or realive to the current working directory:

     system$ re.pl --rcfile /path/to/my/project/repl.rc

    Within the run control file you might want to load plugins. This is
    covered in "The REPL shell object" section, below.

  Profiles
    To allow for the sharing of run control files, you can fashion them into
    a Perl module for distribution (perhaps via the CPAN). For more
    information on this feature, please see the Devel::REPL::Profile manual
    page.

    A default profile ships with "Devel::REPL"; it loads the following
    plugins:

    *   Devel::REPL::Plugin::History

    *   Devel::REPL::Plugin::LexEnv

    *   Devel::REPL::Plugin::DDS

    *   Devel::REPL::Plugin::Packages

    *   Devel::REPL::Plugin::Commands

    *   Devel::REPL::Plugin::MultiLine::PPI

  Plugins
    Plugins are a way to add funcionality to the REPL shell, and take
    advantage of "Devel::REPL" being based on the Moose object system for
    Perl 5. This means it's simple to 'hook into' many steps of the R-E-P-L
    process. Plugins can change the way commands are interpreted, or the way
    their results are output, or even add commands to the shell environment.

    A number of plugins ship with "Devel::REPL", and more are available on
    the CPAN. Some of the shipped plugins are loaded in the default profile,
    mentioned above.

    Writing your own plugins is not difficult, and is discussed in the
    Devel::REPL::Plugin manual page, along with links to the manual pages of
    all the plugins shipped with "Devel::REPL".

  The REPL shell object
    From time to time you'll want to interact with or manipulate the
    "Devel::REPL" shell object itself; that is, the instance of the shell
    you're currently running.

    The object is always available through the $_REPL variable. One common
    requirement is to load an additional plugin, after your profile and run
    control files have already been executed:

     $_ $_REPL->load_plugin('Timing');
     1
     $_ print "Hello again, world!\n"
     Hello again, world!
     Took 0.00148296356201172 seconds.
     1
     $_

REQUIREMENTS
    In addition to the contents of the standard Perl distribution, you will
    need the following:

    *   Moose >= 0.74

    *   MooseX::Object::Pluggable >= 0.0009

    *   MooseX::Getopt >= 0.18

    *   MooseX::AttributeHelpers >= 0.16

    *   namespace::clean

    *   File::HomeDir

    *   Task::Weaken

    *   B::Concise

    *   Term::ANSIColor

    *   Devel::Peek

    Optionally, some plugins if installed will require the following
    modules:

    *   PPI

    *   Data::Dump::Streamer

    *   Data::Dumper::Concise

    *   File::Next

    *   Sys::SigAction

    *   B::Keywords

    *   Lexical::Persistence

    *   App::Nopaste

    *   Module::Refresh

AUTHOR
    Matt S Trout - mst (at) shadowcatsystems.co.uk
    (<http://www.shadowcatsystems.co.uk/>)

CONTRIBUTORS
    Stevan Little - stevan (at) iinteractive.com
    Alexis Sukrieh - sukria+perl (at) sukria.net
    epitaph
    mgrimes - mgrimes (at) cpan dot org
    Shawn M Moore - sartak (at) gmail.com
    Oliver Gorwits - oliver on irc.perl.org
    Andrew Moore - "<amoore@cpan.org>"
    Norbert Buchmuller "<norbi@nix.hu>"
    Dave Houston "<dhouston@cpan.org>"
    Chris Marshall

LICENSE
    This library is free software under the same terms as perl itself



Changes
1.003012
  - Remove some old code to avoid a future Moose warning
1.003011
  - History plugin now works for Term::ReadLine::Perl
  - Add ReadLine file completion if no other matches, fixes rt.cpan#58351
  - Add a bit more description of Turtle plugin to its POD
  - Fix rt.cpan#43151 where _-> completion had error
  - Add some missing functionality to t/load_core.t test coverage
  - Added docs on quoting and history expansion addressing rt.cpan#48459
1.3.10
  - Add exit_repl attribute as a non-signal related way to indicate the REPL should stop running.
1.3.9_02
  - resolve rt.cpan#45879 by implementing Term::ReadLine::Perl support for ReadLineHistory
  - resolve rt.cpan#55491 by skipping Interrupt test for win32
1.3.9_01
  - Completion works with Term::ReadLine::Perl, which works in win32 etc
1.3.9
  - Fix scary Moose 0.98 box, adding a check for ->isa('Devel::REPL') in Meta/Plugin.pm
1.3.8
  - Add doc to a bunch of plugins
  - Add DDC plugin (like DDS but Data::Dumper::Concise)
  - resolve rt.cpan#44231 SIGINT does not quit the REPL just kills the line being edited
1.3.7
  - resolve rt.cpan#44919 fix deprecated use of compute_all_applicable_methods
1.3.6
  - had to cut a new release, something went wrong with use of M::I
1.3.5
  - resolve rt.cpan#43807 three Completion drivers not loading Completion plugin
  - resolve rt.cpan#43808 infinite recursion on Plugin::Commands when Plugin::Packages not loaded
  - resolve rt.cpan#44902 warning due to deprecated use of alias_method in Plugin.pm
  - add comment formatting option to the nopaste plugin
  - add #pastetitle command to Nopaste plugin
  - If the object we're dumping has stringification overloading, then use
    it directly
1.3.4
  - resolve rt.cpan#43109 Missing deps for Default Profile Plugins
1.3.3
  - make Nopaste attribute lazy
  - make ReadLineHistory history file location portable
  - resolve rt.cpan#42904 Nopaste plugin handle undef error messages
1.3.2
  - make plugins be M::I features so dependencies are optional
  - resolve rt.cpan#42506 readdir might fail
  - test use of all plugins supported by Perl core
  - Add irc nick for oliver
1.3.1
  - Fix brokenness in the distribution manifest
1.3.0
  - OutputCache: If sub _ is already defined, then warn about it (once) and refuse to overwrite it
  - Turn MultiLine::PPI's needs_continuation into a method line_needs_continuation
  - Cleanup Plugin::Packages
  - Print a warning when Completion is loaded but the Term::ReadLine object doesn't support completion
  - Introduce error object for error_return
  - Make compile return a single value, use is_error to test if it's bad (it's now an error object instead of undef + error str)
  - Refactor to make it easier to add print/warn to the session
  - Preserve $^H and %^H when using the LexEnv plugin
  - Refactor the finding of the last element of the document into a method in Completion
  - Use namespaace::clean in turtles completion driver
  - Use Devel::REPL::Plugin instead of Moose::Role
  - Use "BEFORE_PLUGIN { load_plugin }" instead of "with"
  - Fix from Sartak for end-of-line comment bug reported by jrockway
  - Allow exception objects.
  - Don't initialize a turtles matcher if we don't need to
  - Improve the MultiLine::PPI plugin by checking whether adding ;; will end in PPI::Statement::Null. If not, then there must be some incomplete structure
  - Respect env var DEVEL_REPL_PROFILE
  -
  - Devel::REPL::Plugin::Peek
  - Devel::REPL::Plugin::B::Concise
  - Add CompletionDriver::Turtles
  - Devel::REPL::Error
  - Add a Nopaste plugin. #nopaste will publish your current session using App::Nopaste
  - Add the @INC completion driver.
  - Devel::REPL::Plugin::CompletionDriver::Globals
  - Add a PPI plugin for #ppi CODE
  - CompletionDriver::Methods which currently only works on classnames
  - FindVariable plugin, which returns a reference to a global variable by name
  -
  - Update dependencies
  - Give "NAME" pod sections to all the plugins
  - Add AUTHOR sections where I know off the top of my head who wrote the plugin
  - Expand some of the user documentation
1.2.1
  - ship a working version of 1.2.0
1.2.0
  - fix Plugin to work with latest Moose
  - deprecate NewlineHack plugin, cause fixed in REPL.pm
  - OutputCache plugin
  - MultiLine::PPI plugin
  - Timing plugin
  - Interrupt plugin
  - FancyPrompt plugin
  - Colors plugin
1.1.0
  - profile system
  - fix up dependencies
  - DDS plugin
  - Commands plugin
  - Plugin metaclass
1.0.0
  - Initial release


lib/Devel/REPL.pm
package Devel::REPL;

use Term::ReadLine;
use Moose;
use namespace::clean -except => [ 'meta' ];
use 5.008001; # backwards compat, doesn't warn like 5.8.1

our $VERSION = '1.003012';

with 'MooseX::Object::Pluggable';

use Devel::REPL::Error;

has 'term' => (
  is => 'rw', required => 1,
  default => sub { Term::ReadLine->new('Perl REPL') }
);

has 'prompt' => (
  is => 'rw', required => 1,
  default => sub { '$ ' }
);

has 'out_fh' => (
  is => 'rw', required => 1, lazy => 1,
  default => sub { shift->term->OUT || \*STDOUT; }
);

has 'exit_repl' => (
  is => 'rw', required => 1,
  default => sub { 0 }
);

sub run {
  my ($self) = @_;
  while ($self->run_once_safely) {
    # keep looping unless we want to exit REPL
    last if $self->exit_repl;
  }
}

sub run_once_safely {
  my ($self, @args) = @_;

  my $ret = eval { $self->run_once(@args) };

  if ($@) {
    my $error = $@;
    eval { $self->print("Error! - $error\n"); };
    return 1;
  } else {
    return $ret;
  }
}

sub run_once {
  my ($self) = @_;

  my $line = $self->read;
  return unless defined($line);     # undefined value == EOF

  my @ret = $self->formatted_eval($line);

  $self->print(@ret) unless $self->exit_repl;

  return 1;
}

sub formatted_eval {
  my ( $self, @args ) = @_;

  my @ret = $self->eval(@args);

  return $self->format(@ret);
}

sub format {
  my ( $self, @stuff ) = @_;

  if ( $self->is_error($stuff[0]) ) {
    return $self->format_error(@stuff);
  } else {
    return $self->format_result(@stuff);
  }
}

sub format_result {
  my ( $self, @stuff ) = @_;

  return @stuff;
}

sub format_error {
  my ( $self, $error ) = @_;
  return $error->stringify;
}

sub is_error {
  my ( $self, $thingy ) = @_;
  blessed($thingy) and $thingy->isa("Devel::REPL::Error");
}

sub read {
  my ($self) = @_;
  return $self->term->readline($self->prompt);
}

sub eval {
  my ($self, $line) = @_;
  my $compiled = $self->compile($line);
  return $compiled unless defined($compiled) and not $self->is_error($compiled);
  return $self->execute($compiled);
}

sub compile {
  my ( $_REPL, @args ) = @_;
  my $compiled = eval $_REPL->wrap_as_sub(@args);
  return $_REPL->error_return("Compile error", $@) if $@;
  return $compiled;
}

sub wrap_as_sub {
  my ($self, $line, %args) = @_;
  return qq!sub {\n!. ( $args{no_mangling} ? $line : $self->mangle_line($line) ).qq!\n}\n!;
}

sub mangle_line {
  my ($self, $line) = @_;
  return $line;
}

sub execute {
  my ($self, $to_exec, @args) = @_;
  my @ret = eval { $to_exec->(@args) };
  return $self->error_return("Runtime error", $@) if $@;
  return @ret;
}

sub error_return {
  my ($self, $type, $error) = @_;
  return Devel::REPL::Error->new( type => $type, message => $error );
}

sub print {
  my ($self, @ret) = @_;
  my $fh = $self->out_fh;
  no warnings 'uninitialized';
  print $fh "@ret";
  print $fh "\n" if $self->term->ReadLine =~ /Gnu/;
}

=head1 NAME

Devel::REPL - a modern perl interactive shell

=head1 SYNOPSIS

  my $repl = Devel::REPL->new;
  $repl->load_plugin($_) for qw(History LexEnv);
  $repl->run

Alternatively, use the 're.pl' script installed with the distribution

  system$ re.pl

=head1 DESCRIPTION

This is an interactive shell for Perl, commonly known as a REPL - Read,
Evaluate, Print, Loop. The shell provides for rapid development or testing
of code without the need to create a temporary source code file.

Through a plugin system, many features are available on demand. You can also
tailor the environment through the use of profiles and run control files, for
example to pre-load certain Perl modules when working on a particular project.

=head1 USAGE

To start a shell, follow one of the examples in the L</"SYNOPSIS"> above.

Once running, the shell accepts and will attempt to execute any code given. If
the code executes successfully you'll be shown the result, otherwise an error
message will be returned. Here are a few examples:

 $_ print "Hello, world!\n"
 Hello, world!
 1
 $_ nosuchfunction
 Compile error: Bareword "nosuchfunction" not allowed while "strict subs" in use at (eval 130) line 5.

 $_

In the first example above you see the output of the command (C<Hello,
world!>), if any, and then the return value of the statement (C<1>). Following
that example, an error is returned when the execution of some code fails.

Note that the lack of semicolon on the end is not a mistake - the code is
run inside a Block structure (to protect the REPL in case the code blows up),
which means a single statement doesn't require the semicolon. You can add one
if you like, though.

If you followed the first example in the L</"SYNOPSIS"> above, you'll have the
History and LexEnv plugins loaded (and there are many more available).
Although the shell might support "up-arrow" history, the History plugin adds
"bang" history to that so you can re-execute chosen commands (with e.g.
C<!53>). The LexEnv plugin ensures that lexical variables declared with the
C<my> keyword will automatically persist between statements executed in the
REPL shell.

When you C<use> any Perl module, the C<import()> will work as expected - the
exported functions from that module are available for immediate use:

 $_ carp "I'm dieeeing!\n"
 String found where operator expected at (eval 129) line 5, near "carp "I'm dieeeing!\n""
         (Do you need to predeclare carp?)
 Compile error: syntax error at (eval 129) line 5, near "carp "I'm dieeeing!\n""
 BEGIN not safe after errors--compilation aborted at (eval 129) line 5.

 $_ use Carp

 $_ carp "I'm dieeeing!\n"
 I'm dieeeing!
  at /usr/share/perl5/Lexical/Persistence.pm line 327
 1
 $_

To quit from the shell, hit C<Ctrl+D> or C<Ctrl+C>.

  MSWin32 NOTE: control keys won't work if TERM=dumb
  because readline functionality will be disabled.


=head2 Run Control Files

For particular projects you might well end up running the same commands each
time the REPL shell starts up - loading Perl modules, setting configuration,
and so on. A run control file lets you have this done automatically, and you
can have multiple files for different projects.

By default the C<re.pl> program looks for C<< $HOME/.re.pl/repl.rc >>, and
runs whatever code is in there as if you had entered it at the REPL shell
yourself.

To set a new run control file that's also in that directory, pass it as a
filename like so:

 system$ re.pl --rcfile myproject.pc

If the filename happens to contain a forwardslash, then it's used absolutely,
or realive to the current working directory:

 system$ re.pl --rcfile /path/to/my/project/repl.rc

Within the run control file you might want to load plugins. This is covered in
L</"The REPL shell object"> section, below.

=head2 Profiles

To allow for the sharing of run control files, you can fashion them into a
Perl module for distribution (perhaps via the CPAN). For more information on
this feature, please see the L<Devel::REPL::Profile> manual page.

A default profile ships with C<Devel::REPL>; it loads the following plugins:

=over 4

=item *

L<Devel::REPL::Plugin::History>

=item *

L<Devel::REPL::Plugin::LexEnv>

=item *

L<Devel::REPL::Plugin::DDS>

=item *

L<Devel::REPL::Plugin::Packages>

=item *

L<Devel::REPL::Plugin::Commands>

=item *

L<Devel::REPL::Plugin::MultiLine::PPI>

=back

=head2 Plugins

Plugins are a way to add funcionality to the REPL shell, and take advantage of
C<Devel::REPL> being based on the L<Moose> object system for Perl 5. This
means it's simple to 'hook into' many steps of the R-E-P-L process. Plugins
can change the way commands are interpreted, or the way their results are
output, or even add commands to the shell environment.

A number of plugins ship with C<Devel::REPL>, and more are available on the
CPAN. Some of the shipped plugins are loaded in the default profile, mentioned
above.

Writing your own plugins is not difficult, and is discussed in the
L<Devel::REPL::Plugin> manual page, along with links to the manual pages of
all the plugins shipped with C<Devel::REPL>.

=head2 The REPL shell object

From time to time you'll want to interact with or manipulate the
C<Devel::REPL> shell object itself; that is, the instance of the shell you're
currently running.

The object is always available through the C<$_REPL> variable. One common
requirement is to load an additional plugin, after your profile and run
control files have already been executed:

 $_ $_REPL->load_plugin('Timing');
 1
 $_ print "Hello again, world!\n"
 Hello again, world!
 Took 0.00148296356201172 seconds.
 1
 $_

=head1 REQUIREMENTS

In addition to the contents of the standard Perl distribution, you will need
the following:

=over 4

=item *

L<Moose> >= 0.74

=item *

L<MooseX::Object::Pluggable> >= 0.0009

=item *

L<MooseX::Getopt> >= 0.18

=item *

L<MooseX::AttributeHelpers> >= 0.16

=item *

L<namespace::clean>

=item *

L<File::HomeDir>

=item *

L<Task::Weaken>

=item *

L<B::Concise>

=item *

L<Term::ANSIColor>

=item *

L<Devel::Peek>

=back

Optionally, some plugins if installed will require the following modules:

=over 4

=item *

L<PPI>

=item *

L<Data::Dump::Streamer>

=item *

L<Data::Dumper::Concise>

=item *

L<File::Next>

=item *

L<Sys::SigAction>

=item *

L<B::Keywords>

=item *

L<Lexical::Persistence>

=item *

L<App::Nopaste>

=item *

L<Module::Refresh>

=back

=head1 AUTHOR

Matt S Trout - mst (at) shadowcatsystems.co.uk (L<http://www.shadowcatsystems.co.uk/>)

=head1 CONTRIBUTORS

=over 4

=item Stevan Little - stevan (at) iinteractive.com

=item Alexis Sukrieh - sukria+perl (at) sukria.net

=item epitaph

=item mgrimes - mgrimes (at) cpan dot org

=item Shawn M Moore - sartak (at) gmail.com

=item Oliver Gorwits - oliver on irc.perl.org

=item Andrew Moore - C<< <amoore@cpan.org> >>

=item Norbert Buchmuller C<< <norbi@nix.hu> >>

=item Dave Houston C<< <dhouston@cpan.org> >>

=item Chris Marshall

=back

=head1 LICENSE

This library is free software under the same terms as perl itself

=cut

1;
